File: src/package.ts
import { RequestHandler } from 'express';
import { maxSatisfying } from 'semver';
import got from 'got';
import { NPMPackage } from './types';

type Package = { version: string; dependencies: Record<string, Package> };
// idea: since npm has multiple packages to traverse through, it might be worth
// adding cache. we can also add TTL to cache, such as npmCache
// review: 

/**
 * Attempts to retrieve package data from the npm registry and return it
 */
export const getPackage: RequestHandler = async function (req, res, next) {
  const { name, version } = req.params;
  const dependencyTree = {};
  try {
    const npmPackage: NPMPackage = await got(
      `https://registry.npmjs.org/${name}`,
    ).json();

    const dependencies: Record<string, string> =
      npmPackage.versions[version].dependencies ?? {};
    for (const [name, range] of Object.entries(dependencies)) {
      const subDep = await getDependencies(name, range);
      dependencyTree[name] = subDep;
    }
    // review: 
    return res
      .status(200)
      .json({ name, version, dependencies: dependencyTree });
  } catch (error) {
    // review: 
    return next(error);
  }
};

async function getDependencies(name: string, range: string): Promise<Package> {
  // review: 
  const npmPackage: NPMPackage = await got(
    `https://registry.npmjs.org/${name}`,
  ).json();

  const v = maxSatisfying(Object.keys(npmPackage.versions), range);
  // review: 
  const dependencies: Record<string, Package> = {};
  // idea: async or batch calls 
  if (v) {
    const newDeps = npmPackage.versions[v].dependencies;
    for (const [name, range] of Object.entries(newDeps ?? {})) {
      dependencies[name] = await getDependencies(name, range);
    }
  }

  return { version: v ?? range, dependencies };
}

File: src/package.py
from typing import Optional

import requests
from semver import max_satisfying

from src.models import NPMPackage, NPMPackageVersion

NPM_REGISTRY_URL = "https://registry.npmjs.org"


async def get_package(name: str, version: Optional[str] = None) -> NPMPackageVersion:
    url = f"{NPM_REGISTRY_URL}/{name}"
    npm_package = requests.get(url).json()
    package = NPMPackage(
        name=npm_package["name"],
        description=npm_package["description"],
        dist_tags=npm_package["dist-tags"],
        versions=npm_package["versions"],
    )
    if not version:
        version = max_satisfying(package.versions.keys(), "*")

    dependencies = package.versions[version].dependencies
    return NPMPackageVersion(name=name, version=version, dependencies=dependencies)
